Adding transpose node and changes to normalize node

SundarRajan28 opened the pull request.

SundarRajan28 committed:
Adding transpose node and changes to normalize node

fiona-gladwin commented on rocAL/include/api/rocal_api_augmentation.h:
```diff
@@ -354,6 +354,20 @@ extern "C" RocalTensor ROCAL_API_CALL rocalFlipFixed(RocalContext context, Rocal
+ * \return RocalTensor
+ */
+extern "C" RocalTensor ROCAL_API_CALL rocalTranspose(RocalContext context, RocalTensor input, std::vector<unsigned> perm, bool is_output,
+                                                RocalTensorLayout output_layout = ROCAL_NONE,
```
Align the arguments

fiona-gladwin commented on rocAL/include/augmentations/geometry_augmentations/node_transpose.h:
```diff
@@ -0,0 +1,40 @@
+
+   private:
+    std::vector<unsigned> _perm;
+    vx_array _perm_array;
```
Can we name it _perm_array_vx

fiona-gladwin commented on rocAL/source/augmentations/geometry_augmentations/node_transpose.cpp:
```diff
@@ -0,0 +1,51 @@
+    vx_scalar roi_type_vx = vxCreateScalar(vxGetContext((vx_reference)_graph->get()), VX_TYPE_INT32, &roi_type);
+    _perm_array = vxCreateArray(vxGetContext((vx_reference)_graph->get()), VX_TYPE_UINT32, _perm.size());
+    vx_status status = VX_SUCCESS;
+    status |= vxAddArrayItems(_perm_array, _perm.size(), _perm.data(), sizeof(vx_uint32));
```
Why status |= is used?

SundarRajan28 replied:
|= is used in all the vxAddArrayItems instances in other nodes

fiona-gladwin commented on rocAL/source/api/rocal_api_augmentation.cpp:
```diff
@@ -1200,6 +1200,37 @@ rocalSnPNoiseFixed(
+        RocalTensorDataType op_tensor_datatype = static_cast<RocalTensorDataType>(output_datatype);
+        TensorInfo output_info = input->info();
+        output_info.set_tensor_layout(op_tensor_layout);
+        output_info.set_data_type(op_tensor_datatype);
```
The ROI will change with this transpose and it will be different from the input ROI right?
In that case we need to reset the ROI so call a set max shape here.

Check with cascaded augmentations in python end.

fiona-gladwin commented on rocAL/source/augmentations/effects_augmentations/node_normalize.cpp:
```diff
@@ -43,14 +43,42 @@ void NormalizeNode::create_node() {
-            totalElements *= !axis[j] ? tensor_shape[j] : 1;
+    auto data_layout = _inputs[0]->info().layout();
+    auto tensor_dims = _inputs[0]->info().dims();
+    if (data_layout == RocalTensorlayout::NHWC) {
```
Why is this done?
Please add suitable comments and can we make it generic?

SundarRajan28 replied:
Only in NHWC and NCHW layouts we store the ROI in XYWH format. For other formats, we have the generic ROI which is handled by the else case

fiona-gladwin commented on rocAL_pybind/amd/rocal/fn.py:
```diff
@@ -1229,3 +1229,9 @@ def mel_filter_bank(*inputs, bytes_per_sample_hint = [0], freq_high = 0.0, freq_
     return mel_filter_bank_output
+
+def transpose(*inputs, perm=[], output_layout=types.NHWC, output_dtype=types.UINT8):
+    # pybind call arguments
```
Add docstring similar to others

fiona-gladwin replied:
Also explain the arguments

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Minor changes

SundarRajan28 committed:
Add ROI updation in transpose node and test case addition

SundarRajan28 committed:
Update test case

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into transpose_normalize_support

SundarRajan28 committed:
Merge remote-tracking branch 'origin/numpy_reader_python_changes' into transpose_normalize_support

SundarRajan28 committed:
Numpy reader - python changes (#292)

* Revert empty line removed in CMakeLists.txt

* Removed prefix original for audio vectors

* Fix PARTIAL

Fix issue with batch size greater than dataset size

* Reduce overall time for audio source evalution

* Fix shard_size and stick to shard issue seen with convergence

* Resolve PR comments

* Add @params to all args in pytorch.py

* Fix build issue

* Minor changes in unit test

* Minor changes

* Change ROCAL instaces to rocAL in pytorch.py

* Resolve the PR comments

* Minor changes in decoders.py - Modify the comment for shard_size

* Fix shard_size

* Minor changes

* Changes in pipeline.py and decoders.py

* Address the PR comments

* Address Review comments

* Remove print statement

* Fix the count_items

* Make Sharding similar to DALI

* Fix issues with DROP policy by introducing a new vector for padding

* Minor fixes

* Comment out print statements

* Add changes for shard_size LBP testing

* Fix DROP Policy with shard_size > 0

* Fix Stick_to_Shard=False

* Fix PARTIAL policy and code clean up

* fix last_batch_padded size when shard_size > 0

* Fix Drop policy - we skip the dropped batch in the next epoch

* Fix single shard outputs

* Remove the commented code and fix the padding code in open()

* Remove div by num_shards in decoders.py

* Introduce Audio layouts

* Add layout changes for spectrogram

* Fix the unit tests - c++ & python

* Code clean up and formatting

* Minor code clean up

* code clean up in pytorch.py

* Add layout changes for spectrogram

* Pass layouts for MelFilterBank

* Fix ToDecibels

Pass layouts for ToDecibels

* Fix Normalize

* Fix build issue

* Fix python unit test

* Minor fix

* Pass LBP to decoders instead of the Pipeline creation

* Update pipeline.py - Remove commented code

* Update pipeline.py - Remove commented out code

* Adding changes for spec layout changes

* Adding changes to MFB and normalize nodes

* Update node_slice.cpp

* Update node_slice.h

* Resolve PR comments

* Fix downmix failing case and resolve the issue with merge

* Fix issue with file_source_reader.cpp when file_list is not used

* Add the changes for LBP in coco-file-source-reader.

* Resolve PR comments - Sundar

* Fix undefined reference

* Fix file_source_reader.cpp

* Fix shuffle issues

* Fix coco_file_source

* make stick_to_shard as true by default

* Fix minor bug

* Change last_patch_padded to pad_last_batch_repeated

* Fix the shuffle with aspect ratios

* Make a minor change in the coco-file-source - sorting is not required

* Another fix for shuffle with aspect_ratios

* Another fix for shuffle with aspect_ratios

* Modify the split_count

* Fix name of shard_count

* Adding print statements

* Add the sorting back

* Fix the indexs calc

* Fix the indexs calc

* Add some debug statements

* Fix the ARG

* comment out part of ASR

* Comment out shuffle_with_asr

* Uncomment Line 234 & 236

* Uncomment Line 234 & 236

* Uncomment the ARG

* Add Cmake changes for HIP backend

* Add support for TF record reader

* Adding changes for lbp & sharding in caffe reader

* Adding changes for lbp & sharding in caffe2 lmdb reader

* Adding changes for lbp & sharding in cifar10 reader

* Adding changes for lbp & sharding in mxnetIO reader

* Minor change in tf record reader

* Adding partial LBP changes to np reader

* Adding comments to all if conditions

* Adding the lbps function implementation

* Minor change

* Fix a bug in the caffe2 reader

* Minor changes

* Fix count_items

* Fix merge conflicts

* Resolving review comments

* Resolving review comments

* Fix a minor warning in file source reader

* Resolving review comments

* LBP comments resolution

* Resolving review comments

* Formatting changes

* Resolving Final Set of PR comments

* Combine with OR condition

* Remove the pad_last_batch_repeated print statement from decoders.py

* Add shard_size and stick_to_shard variables in args

* Minor spelling fix

* Make changes to insert the padded data in the file_names vector

* Support to pass the variables fo lbp as struct

* Fix segmentation fault

* Removing PR11 changes

* Adding LBP changes to numpy reader

* Adding i16 support to rocAL and np reader

* Add npy version check

* Resolving review comments

* Adding numpy loader changes

* Change copyright year

* Resolving review comments

* Resolve review comments

* Fixing issue with setting dtype

* Minor changes

* Adding ROI changes for numpy reader

* Updating unit test for numpy reader

* Adding python changes for numpy reader

* Resolving review comments

* Adding function comments

* Adding checks for numpy header parsing

* Adding fixes for checks

* Adding reader test for np reader

* Resolving review comments

* Adding changes from PR267

* Resolving review comments

* Resolving review comments

* Resolving review comments

* Macro changes

* Resolving review comments

* Macro changes

* Removing templated function for reading numpy data

* Resolving review comments

* Minor changes

* Minor changes

* Update changelog

* Adding reader test list in readme

---------

Co-authored-by: fiona-gladwin <fionagladwin@multicorewareinc.com>
Co-authored-by: Swwetha B S <swetha@multicorewareinc.com>
Co-authored-by: root <root@x1000c4s5b0n0.hostmgmt2000.cm.lockhart.amd.com>
Co-authored-by: swetha097 <59434434+swetha097@users.noreply.github.com>
Co-authored-by: SundarRajan98 <svaithiy@amd.com>
Co-authored-by: sbavasab <102275787+sbavasab@users.noreply.github.com>
Co-authored-by: Lakshmi Kumar <lakshmi.kumar@amd.com>
Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

kiritigowda committed:
Merge branch 'develop' into transpose_normalize_support

fiona-gladwin committed:
rocDecode integration: CMakeLists version check fix (#306)

* Update Version check for rocJpeg and rocDecode

* Fix version check

* Modify circular buffer init for rocDecode

---------

Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

kiritigowda committed:
Merge branch 'develop' into transpose_normalize_support

LakshmiKumar23 committed:
Test - Remove data files & readme updates (#305)

* removing video files

* remove all data

* update readme

---------

Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

kiritigowda committed:
Merge branch 'develop' into transpose_normalize_support

danielsu-amd committed:
External CI: enable mainline branch trigger (#308)

kiritigowda committed:
Merge branch 'develop' into transpose_normalize_support

SundarRajan28 committed:
Update normalize code
